src/main/java/com/linecorp/armeria/server/ServiceConfig.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright 2015 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.ServiceInvocationContext;
+
+/**
+ * A {@link Service} and its {@link PathMapping} and {@link VirtualHost}.
+ *
+ * @see ServerConfig#serviceConfigs()
+ * @see VirtualHost#serviceConfigs()
+ */
+public final class ServiceConfig {
+
+    private static final Pattern LOGGER_NAME_PATTERN =
+            Pattern.compile("^\\p{javaJavaIdentifierStart}\\p{javaJavaIdentifierPart}*" +
+                            "(?:\\.\\p{javaJavaIdentifierStart}\\p{javaJavaIdentifierPart}*)*$");
+
+    /** Initialized later by {@link VirtualHost} via {@link #build(VirtualHost)}. */
+    private VirtualHost virtualHost;
+
+    private final PathMapping pathMapping;
+    private final String loggerName;
+    private final Service service;
+
+    private String fullLoggerName;
+
+    /**
+     * Creates a new instance.
+     */
+    public ServiceConfig(VirtualHost virtualHost, PathMapping pathMapping, Service service) {
+        this(virtualHost, pathMapping, service, null);
+    }
+
+    /**
+     * Creates a new instance.
+     */
+    public ServiceConfig(VirtualHost virtualHost, PathMapping pathMapping, Service service,
```

Comment:
Added the constraints of logger names and prefixes to ServerBuilder.service()/serviceLoggerPrefix() and VirtualHostBuilder.service()